python/private/py_library.bzl

```diff
@@ -144,6 +214,73 @@ Source files are no longer added to the runfiles directly.
 :::
 """
 
+def _get_site_packages_symlinks(ctx):
+    if not ctx.attr.site_packages_root:
+        return []
+
+    # We have to build a list of (runfiles path, site-packages path) pairs of
+    # the files to create in the consuming binary's venv site-packages directory.
+    # To minimize the number of files to create, we just return the paths
+    # to the directories containing the code of interest.
+    #
+    # However, namespace packages complicate matters: multiple
+    # distributions install in the same directory in site-packages. This
+    # works out because they don't overlap in their files. Typically, they
+    # install to different directories within the namespace package
+    # directory. Namespace package directories are simply directories
+    # within site-packages that *don't* have an `__init__.py` file, which
+    # can be arbitrarily deep. Thus, we simply have to look for the
+    # directories that _do_ have an `__init__.py` file and treat those as
+    # the path to symlink to.
+
+    site_packages_root = paths.join(ctx.label.package, ctx.attr.site_packages_root)
+    repo_runfiles_dirname = None
+    dirs_with_init = {}  # dirname -> runfile path
+    for src in ctx.files.srcs:
+        if src.extension not in PYTHON_FILE_EXTENSIONS:
+            continue
+        path = _repo_relative_short_path(src.short_path)
+        if not path.startswith(site_packages_root):
+            continue
+        path = path.removeprefix(site_packages_root)
+        dir_name, _, filename = path.rpartition("/")
+        if not dir_name:
+            # This would be e.g. `site-packages/__init__.py`, which isn't valid.
+            # Apparently, the pypi integration adds such a file?
+            continue
+
+        if filename.startswith("__init__."):
+            dirs_with_init[dir_name] = None
+            repo_runfiles_dirname = runfiles_root_path(ctx, src.short_path).partition("/")[0]
+
+    # Sort so that we encounter `foo` before `foo/bar`. This ensures we
+    # see the top-most explicit package first.
+    dirnames = sorted(dirs_with_init.keys())
+    first_level_explicit_packages = []
+    for d in dirnames:
+        is_sub_package = False
+        for existing in first_level_explicit_packages:
+            # Suffix with / to prevent foo matching foobar
+            if d.startswith(existing + "/"):
+                is_sub_package = True
+                break
+        if not is_sub_package:
+            first_level_explicit_packages.append(d)
+
+    site_packages_symlinks = []
+    for dirname in first_level_explicit_packages:
+        site_packages_symlinks.append((
+            paths.join(repo_runfiles_dirname, site_packages_root, dirname),
+            dirname,
+        ))
+    return site_packages_symlinks
+
+def _repo_relative_short_path(short_path):
+    if short_path.startswith("../"):
```

Comment:
Could you add a comment showing an example of `short_path` here? Is it `../+pypi+foo/site-packages`?